// react-tooltip-lite/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactMinimalistPortal = require('react-minimalist-portal');

var _reactMinimalistPortal2 = _interopRequireDefault(_reactMinimalistPortal);

var _position = require('./position');

var _position2 = _interopRequireDefault(_position);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @class Tooltip
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @description A lightweight and responsive tooltip.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


// default colors
var defaultColor = '#fff';
var defaultBg = '#333';

var resizeThrottle = 100;
var resizeThreshold = 5;

var stopProp = function stopProp(e) {
  return e.stopPropagation();
};

var Tooltip = function (_React$Component) {
  _inherits(Tooltip, _React$Component);

  _createClass(Tooltip, null, [{
    key: 'getDerivedStateFromProps',
    value: function getDerivedStateFromProps(nextProps) {
      return nextProps.isOpen ? { hasBeenShown: true } : null;
    }
  }]);

  function Tooltip() {
    _classCallCheck(this, Tooltip);

    var _this = _possibleConstructorReturn(this, (Tooltip.__proto__ || Object.getPrototypeOf(Tooltip)).call(this));

    _this.debounceTimeout = false;
    _this.hoverTimeout = false;


    _this.state = { showTip: false, hasHover: false, ignoreShow: false, hasBeenShown: false };

    _this.showTip = _this.showTip.bind(_this);
    _this.hideTip = _this.hideTip.bind(_this);
    _this.checkHover = _this.checkHover.bind(_this);
    _this.toggleTip = _this.toggleTip.bind(_this);
    _this.startHover = _this.startHover.bind(_this);
    _this.endHover = _this.endHover.bind(_this);
    _this.listenResizeScroll = _this.listenResizeScroll.bind(_this);
    _this.handleResizeScroll = _this.handleResizeScroll.bind(_this);
    return _this;
  }

  _createClass(Tooltip, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      // if the isOpen prop is passed on first render we need to immediately trigger a second render,
      // because the tip ref is needed to calculate the position
      if (this.props.isOpen) {
        // eslint-disable-next-line react/no-did-mount-set-state
        this.setState({ isOpen: true });
      }

      window.addEventListener('resize', this.listenResizeScroll);
      window.addEventListener('scroll', this.listenResizeScroll);
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate(_, prevState) {
      // older versions of react won't leverage getDerivedStateFromProps, TODO: remove when < 16.3 support is dropped
      if (!this.state.hasBeenShown && this.props.isOpen) {
        // eslint-disable-next-line react/no-did-update-set-state
        this.setState({ hasBeenShown: true });

        return setTimeout(this.showTip, 0);
      }

      // we need to render once to get refs in place, then we can make the calculations on a followup render
      // this only has to happen the first time the tip is shown, and allows us to not render every tip on the page with initial render.
      if (!prevState.hasBeenShown && this.state.hasBeenShown) {
        this.showTip();
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      window.removeEventListener('resize', this.listenResizeScroll);
      window.removeEventListener('scroll', this.listenResizeScroll);
      clearTimeout(this.debounceTimeout);
    }
  }, {
    key: 'listenResizeScroll',
    value: function listenResizeScroll() {
      clearTimeout(this.debounceTimeout);

      this.debounceTimeout = setTimeout(this.handleResizeScroll, resizeThrottle);
    }
  }, {
    key: 'handleResizeScroll',
    value: function handleResizeScroll() {
      if (this.state.showTip) {
        // if we're showing the tip and the resize was actually a signifigant change, then setState to re-render and calculate position
        var clientWidth = Math.round(document.documentElement.clientWidth / resizeThreshold) * resizeThreshold;
        this.setState({ clientWidth: clientWidth });
      }
    }
  }, {
    key: 'toggleTip',
    value: function toggleTip() {
      this.state.showTip ? this.hideTip() : this.showTip();
    }
  }, {
    key: 'showTip',
    value: function showTip() {
      if (!this.state.hasBeenShown) {
        // this will render once, then fire componentDidUpdate, which will show the tip
        return this.setState({ hasBeenShown: true });
      }

      this.setState({ showTip: true });
    }
  }, {
    key: 'hideTip',
    value: function hideTip() {
      this.setState({ hasHover: false });
      this.setState({ showTip: false });
    }
  }, {
    key: 'startHover',
    value: function startHover() {
      if (!this.state.ignoreShow) {
        this.setState({ hasHover: true });

        clearTimeout(this.hoverTimeout);
        this.hoverTimeout = setTimeout(this.checkHover, this.props.hoverDelay);
      }
    }
  }, {
    key: 'endHover',
    value: function endHover() {
      this.setState({ hasHover: false });

      clearTimeout(this.hoverTimeout);
      this.hoverTimeout = setTimeout(this.checkHover, this.props.hoverDelay);
    }
  }, {
    key: 'checkHover',
    value: function checkHover() {
      this.state.hasHover ? this.showTip() : this.hideTip();
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          arrow = _props.arrow,
          arrowSize = _props.arrowSize,
          background = _props.background,
          className = _props.className,
          children = _props.children,
          color = _props.color,
          content = _props.content,
          direction = _props.direction,
          distance = _props.distance,
          eventOff = _props.eventOff,
          eventOn = _props.eventOn,
          eventToggle = _props.eventToggle,
          forceDirection = _props.forceDirection,
          isOpen = _props.isOpen,
          padding = _props.padding,
          styles = _props.styles,
          TagName = _props.tagName,
          tipContentHover = _props.tipContentHover,
          tipContentClassName = _props.tipContentClassName,
          useDefaultStyles = _props.useDefaultStyles,
          useHover = _props.useHover;


      var isControlledByProps = !(typeof isOpen === 'undefined');
      var showTip = isControlledByProps ? isOpen : this.state.showTip;

      var wrapperStyles = _extends({
        position: 'relative'
      }, styles);

      var props = {
        style: wrapperStyles,
        ref: function ref(target) {
          _this2.target = target;
        },
        className: className
      };

      var portalProps = {
        // keep clicks on the tip from closing click controlled tips
        onClick: stopProp
      };

      // event handling
      if (eventOff) {
        props[eventOff] = this.hideTip;
      }

      if (eventOn) {
        props[eventOn] = this.showTip;
      }

      if (eventToggle) {
        props[eventToggle] = this.toggleTip;

        // only use hover if they don't have a toggle event
      } else if (useHover && !isControlledByProps) {
        props.onMouseOver = this.startHover;
        props.onMouseOut = tipContentHover ? this.endHover : this.hideTip;
        props.onTouchStart = this.toggleTip;

        if (tipContentHover) {
          portalProps.onMouseOver = this.startHover;
          portalProps.onMouseOut = this.endHover;
          portalProps.onTouchStart = stopProp;
        }
      }

      // conditional rendering of tip
      var tipPortal = void 0;

      if (this.state.hasBeenShown) {
        var currentPositions = (0, _position2.default)(direction, forceDirection, this.tip, this.target, _extends({}, this.state, { showTip: showTip }), {
          background: useDefaultStyles ? defaultBg : background,
          arrow: arrow,
          arrowSize: arrowSize,
          distance: distance
        });

        var tipStyles = _extends({}, currentPositions.tip, {
          background: useDefaultStyles ? defaultBg : background,
          color: useDefaultStyles ? defaultColor : color,
          padding: padding,
          boxSizing: 'border-box',
          zIndex: 1000,
          position: 'absolute',
          display: 'inline-block'
        });

        var arrowStyles = _extends({}, currentPositions.arrow, {
          position: 'absolute',
          width: '0px',
          height: '0px',
          zIndex: 1001
        });

        tipPortal = _react2.default.createElement(
          _reactMinimalistPortal2.default,
          null,
          _react2.default.createElement(
            'div',
            _extends({}, portalProps, { className: typeof tipContentClassName !== 'undefined' ? tipContentClassName : className }),
            _react2.default.createElement(
              'span',
              { className: 'react-tooltip-lite', style: tipStyles, ref: function ref(tip) {
                  _this2.tip = tip;
                } },
              content
            ),
            _react2.default.createElement('span', { className: 'react-tooltip-lite-arrow react-tooltip-lite-' + currentPositions.realDirection + '-arrow', style: arrowStyles })
          )
        );
      }

      return _react2.default.createElement(
        TagName,
        props,
        children,
        tipPortal
      );
    }
  }]);

  return Tooltip;
}(_react2.default.Component);

Tooltip.propTypes = {
  arrow: _propTypes2.default.bool,
  arrowSize: _propTypes2.default.number,
  background: _propTypes2.default.string,
  children: _propTypes2.default.node.isRequired,
  className: _propTypes2.default.string,
  color: _propTypes2.default.string,
  content: _propTypes2.default.node.isRequired,
  direction: _propTypes2.default.string,
  distance: _propTypes2.default.number,
  eventOff: _propTypes2.default.string,
  eventOn: _propTypes2.default.string,
  eventToggle: _propTypes2.default.string,
  forceDirection: _propTypes2.default.bool,
  hoverDelay: _propTypes2.default.number,
  isOpen: _propTypes2.default.bool,
  padding: _propTypes2.default.string,
  styles: _propTypes2.default.object,
  tagName: _propTypes2.default.string,
  tipContentHover: _propTypes2.default.bool,
  tipContentClassName: _propTypes2.default.string,
  useDefaultStyles: _propTypes2.default.bool,
  useHover: _propTypes2.default.bool
};
Tooltip.defaultProps = {
  arrow: true,
  arrowSize: 10,
  background: '',
  className: '',
  color: '',
  direction: 'up',
  distance: undefined,
  forceDirection: false,
  hoverDelay: 200,
  padding: '10px',
  styles: {},
  tagName: 'div',
  tipContentHover: false,
  tipContentClassName: undefined,
  useDefaultStyles: false,
  useHover: true
};
exports.default = Tooltip;
